File: obexx/DemoTest/app/src/main/java/com/kenny/demotest/AnimalManager.java
package com.kenny.demotest;

import android.util.Log;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author 29003
 * @description
 * @date 2019/7/23
 */
public class AnimalManager {
    List<State> allAnimalList = null;

    private static AnimalManager instance;
    private AnimalManager (){

    }
    public static synchronized AnimalManager getInstance(){
        if (instance == null){
            instance = new AnimalManager();
        }
        return instance;
    }

    public void setAnimalList(List<State> mutableList){
        this.allAnimalList = mutableList;
    }

    public int getNextAnimalId(int animalId){
        int id = 0;
        String nextState = null;
        String[] next = null;
        if (allAnimalList!=null){
            for(int x=0;x<allAnimalList.size();x++){
                if(allAnimalList.get(x).getWdID() == animalId){
                    nextState = allAnimalList.get(x).getNextState()
                            .replace("\\[", "").replace("\\]", "");
                    break;
                }
            }
            if (nextState!=null){
                next =  nextState.split(",");
            }
            if (next!=null){
                id = getWeightRandom(Arrays.asList(next));
            }
        }
        return id;
    }

    private Double weightArraySum(double[] weightArrays){
        double weightSum = 0.0;
        for (double weightValue :weightArrays) {
            weightSum += weightValue;
        }
        return weightSum;
    }
    /**
     * 根据随机数获取动作ID
     *
     * @param actions
     * @return
     */
    public int getWeightRandom(List<String> actions){
        int action = 0;
        if (actions.size() !=0){
            double[] weightArrays = new double[actions.size()];
            int[] actionIds = new int[actions.size()];
            List<String> ids = new ArrayList<>();
            for (int i=0;i<actions.size();i++){
                for (String s : actions.get(i).split("_")) {
                    if (!s.isEmpty()) {
                        ids.add(s);
                    }
                }
                actionIds[i] = Integer.parseInt(ids.get(0));
                weightArrays[i] = Double.valueOf(ids.get(1));
            }
            if(actionIds.length ==1){
                return actionIds[0];
            }
            Double weightSum = weightArraySum(weightArrays);
            Log.d("随机", "weighsunm==="+weightSum);
            double stepWeightSum=0.0;
            for(int i =0;i<weightArrays.length;i++){
                stepWeightSum += weightArrays[i];
                if (Math.random() <= stepWeightSum / weightSum) {
                    Log.d("随机", "随机下一个动作 i===$i");
                    return actionIds[i];
                }
            }
            Log.d("随机", "随机下一个动作错误===0");
            action = actionIds[0];
        }
        return  action;
    }
}

File: obexx/sdkDemo-master/app/src/main/java/com/obex/demo/model/AnimalManager.kt
package com.obex.demo.model

import android.util.Log

/*
 * -----------------------------------------------------------------
 * Copyright (C) 2014-2016, by obex, All rights reserved.
 * -----------------------------------------------------------------
 *
 * File: AnimalManager.java
 * Author: liuhai
 * Version: V100R001C01
 * Create: 2019/7/11 15:01
 * Changes (from 2019/7/11)
 * -----------------------------------------------------------------
 * 2019/7/11 : Create AnimalManager.java (liuhai);
 * -----------------------------------------------------------------
 * description:
 */

class AnimalManager {
    var allAnimalList: MutableList<AnimalBean>? = null //动作列表

    companion object {
        val instant: AnimalManager by lazy(mode = LazyThreadSafetyMode.SYNCHRONIZED) {
            AnimalManager()
        }
    }

    open fun setAnimalList(mutableList: MutableList<AnimalBean>) {
        this.allAnimalList = mutableList
    }

    open fun getNextAnimalId(animalId: Int): Int {
        var id = 0
        val nextState = allAnimalList?.filter {
            it.wdID == animalId
        }?.first()?.nextState?.replace("\\[", "")?.replace("\\]", "")

        val next = nextState?.split(",")
        next.let {
            id = getWeightRandom(it)
        }
        return id

    }

    private fun weightArraySum(weightArrays: DoubleArray): Double {
        var weightSum = 0.0
        for (weightValue in weightArrays) {
            weightSum += weightValue
        }
        return weightSum
    }

    /**
     * 根据随机数获取动作ID
     *
     * @param actions
     * @return
     */
    fun getWeightRandom(actions: List<String>?): Int {
        var action = 0
        actions?.let {
            val weightArrays = DoubleArray(actions.size)
            val actionIds = IntArray(actions.size)
            for (i in actions.indices) {
                val ids = actions[i].split("_".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                actionIds[i] = Integer.parseInt(ids[0])
                weightArrays[i] = java.lang.Double.valueOf(ids[1])
            }
            if (actionIds.size == 1) {
                return actionIds[0]
            }
            val weightSum = weightArraySum(weightArrays)
            Log.d("随机", "weighsunm===$weightSum")
            var stepWeightSum = 0.0
            for (i in weightArrays.indices) {
                stepWeightSum += weightArrays[i]
                if (Math.random() <= stepWeightSum / weightSum) {
                    Log.d("随机", "随机下一个动作 i===$i")
                    return actionIds[i]
                }
            }
            Log.d("随机", "随机下一个动作错误===0")
            action = actionIds[0]
        }
        return action
    }

}